### PR TITLE
Bug: raise TypeError when doing infer_freq using RangeIndex

### LINKED ISSUE #48439: BUG: `pd.infer_freq()` does not raise an error when called with a `RangeIndex`
## Actual behaviour
Given a `RangeIndex`, the function returns `'N'`, indicating a nanosecond frequency. The function doc clearly states that a `DatetimeIndex` or `TimedeltaIndex` is required as an argument

```python
>> import pandas as pd
>> pd.infer_freq((pd.RangeIndex(1, 10, 1)))
'N'
```

## Expected behavior
A `TypeError` should be raised, probably.

## Possible fix

I believe the issue is in [this code block(https://github.com/pandas-dev/pandas/blob/v1.4.4/pandas/tseries/frequencies.py#L193) -> If we have an non-datetime index, we only raise an error if it's a `Int64Index` or a `Float64Index`

## Output of `pd.show_versions()`
<details>
<summary>Here</summary>

```
INSTALLED VERSIONS
------------------
commit           : ca60aab7340d9989d9428e11a51467658190bb6b
python           : 3.8.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.4.0-125-generic
Version          : #141-Ubuntu SMP Wed Aug 10 13:42:03 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.4
numpy            : 1.23.2
pytz             : 2022.2.1
dateutil         : 2.8.2
setuptools       : 44.0.0
pip              : 20.0.2
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : None
pandas_datareader: None
bs4              : None
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
markupsafe       : None
matplotlib       : None
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
zstandard        : None
```
</details>

### PR DESCRIPTION
**Pull request by ambujpawar**
- [x] closes #48439 
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

File: pandas/tests/tseries/frequencies/test_inference.py [hunk 1]
```diff
@@ -358,7 +359,13 @@ def test_non_datetime_index2():
 
 
 @pytest.mark.parametrize(
-    "idx", [tm.makeIntIndex(10), tm.makeFloatIndex(10), tm.makePeriodIndex(10)]
+    "idx",
+    [
+        tm.makeIntIndex(10),
+        tm.makeFloatIndex(10),
+        tm.makePeriodIndex(10),
+        RangeIndex(10),  # see gh-48439
```
Comment (phofl) [inline pandas/tests/tseries/frequencies/test_inference.py:367]: Could you use tm.makeRangeIndex?

**Review by phofl — CHANGES_REQUESTED** [review]
Could you please add a whatsnew?

File: pandas/tests/tseries/frequencies/test_inference.py (hunk 1 above)
Comment (phofl) [reply pandas/tests/tseries/frequencies/test_inference.py:367]: No gh reference needed here, please add it into the test itself

File: doc/source/whatsnew/v1.6.0.rst [hunk 2]
```diff
@@ -213,6 +213,7 @@ Styler
 
 Other
 ^^^^^
+- Bug in :func:`pandas.infer_freq`, raising ``TypeError`` when inferred on :class:`RangeIndex` (:issue:`47084`)
```
Comment (mroeschke) [inline doc/source/whatsnew/v1.6.0.rst:216]: Could you put this in the `Datetimelike` section?

Comment (ambujpawar) [reply doc/source/whatsnew/v1.6.0.rst]: Done! :)

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
LGTM pending @phofl review.

**Review by phofl — APPROVED** [review]

**Comment by phofl** [comment]
thx @ambujpawar

**Pull request merged**